src/shims/sync.rs

```diff
@@ -168,105 +138,160 @@ fn mutex_get_or_create_blockset<'mir, 'tcx: 'mir>(
-const PTHREAD_RWLOCK_T_MIN_SIZE: u64 = 20;
+const PTHREAD_RWLOCK_T_MIN_SIZE: u64 = 32;
 
-fn rwlock_get_readers<'mir, 'tcx: 'mir>(
+fn rwlock_get_id<'mir, 'tcx: 'mir>(
     ecx: &MiriEvalContext<'mir, 'tcx>,
     rwlock_op: OpTy<'tcx, Tag>,
 ) -> InterpResult<'tcx, ScalarMaybeUndef<Tag>> {
-    get_at_offset(ecx, rwlock_op, 4, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    get_at_offset(ecx, rwlock_op, 12, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
 }
 
-fn rwlock_set_readers<'mir, 'tcx: 'mir>(
+fn rwlock_set_id<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
     rwlock_op: OpTy<'tcx, Tag>,
-    readers: impl Into<ScalarMaybeUndef<Tag>>,
+    id: impl Into<ScalarMaybeUndef<Tag>>,
 ) -> InterpResult<'tcx, ()> {
-    set_at_offset(ecx, rwlock_op, 4, readers, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    set_at_offset(ecx, rwlock_op, 12, id, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
 }
 
-fn rwlock_get_writers<'mir, 'tcx: 'mir>(
-    ecx: &MiriEvalContext<'mir, 'tcx>,
+fn rwlock_get_or_create_id<'mir, 'tcx: 'mir>(
+    ecx: &mut MiriEvalContext<'mir, 'tcx>,
     rwlock_op: OpTy<'tcx, Tag>,
+) -> InterpResult<'tcx, RwLockId> {
+    let id = rwlock_get_id(ecx, rwlock_op)?.to_u32()?;
+    if id == 0 {
+        // 0 is a default value and also not a valid rwlock id. Need to allocate
+        // a new read-write lock.
+        let id = ecx.rwlock_create();
+        rwlock_set_id(ecx, rwlock_op, id.to_u32_scalar())?;
+        Ok(id)
+    } else {
+        Ok(RwLockId::from_u32(id))
+    }
+}
+
+// pthread_condattr_t
+
+// Our chosen memory layout for emulation (does not have to match the platform layout!):
+// store an i32 in the first four bytes equal to the corresponding libc clock id constant
+// (e.g. CLOCK_REALTIME).
+
+const PTHREAD_CONDATTR_T_MIN_SIZE: u64 = 4;
+
+fn condattr_get_clock_id<'mir, 'tcx: 'mir>(
+    ecx: &MiriEvalContext<'mir, 'tcx>,
+    attr_op: OpTy<'tcx, Tag>,
 ) -> InterpResult<'tcx, ScalarMaybeUndef<Tag>> {
-    get_at_offset(ecx, rwlock_op, 8, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    get_at_offset(ecx, attr_op, 0, ecx.machine.layouts.i32, PTHREAD_CONDATTR_T_MIN_SIZE)
 }
 
-fn rwlock_set_writers<'mir, 'tcx: 'mir>(
+fn condattr_set_clock_id<'mir, 'tcx: 'mir>(
     ecx: &mut MiriEvalContext<'mir, 'tcx>,
-    rwlock_op: OpTy<'tcx, Tag>,
-    writers: impl Into<ScalarMaybeUndef<Tag>>,
+    attr_op: OpTy<'tcx, Tag>,
+    clock_id: impl Into<ScalarMaybeUndef<Tag>>,
 ) -> InterpResult<'tcx, ()> {
-    set_at_offset(ecx, rwlock_op, 8, writers, ecx.machine.layouts.u32, PTHREAD_RWLOCK_T_MIN_SIZE)
+    set_at_offset(ecx, attr_op, 0, clock_id, ecx.machine.layouts.i32, PTHREAD_CONDATTR_T_MIN_SIZE)
 }
 
-fn rwlock_get_writer_blockset<'mir, 'tcx: 'mir>(
+// pthread_cond_t
+
+// Our chosen memory layout for the emulated conditional variable (does not have
+// to match the platform layout!):
+
+// bytes 4-7: the conditional variable id as u32 or 0 if id is not assigned yet.
```

Comment:
Is there a specific reason you did not start at offset 0?

Reply:
MacOS stores some stuff in bytes 0-3.

Reply:
The please add a comment explaining that.